crates/autopilot/src/run_loop.rs

```diff
@@ -529,10 +543,39 @@ impl RunLoop {
         solutions
     }
 
+    /// Chooses the winners from the given participants.
+    ///
+    /// Participants are already sorted by their score (best to worst).
+    ///
+    /// Winners are selected one by one, starting from the best solution,
+    /// until `max_winners_per_auction` are selected. The solution is a winner
+    /// if it swaps tokens that are not yet swapped by any other already
+    /// selected winner.
+    fn select_winners<'a>(&self, participants: &'a VecDeque<Participant>) -> Vec<&'a Participant> {
+        let mut winners = Vec::new();
+        let mut already_swapped_tokens = HashSet::new();
+        for participant in participants.iter() {
+            let swapped_tokens = participant
+                .solution
+                .orders()
+                .iter()
+                .map(|(_, order)| (order.sell.token, order.buy.token))
+                .collect::<HashSet<_>>();
```

Comment:
So, we allow solutions with an A->B pair from one winner and B->A from another? Should we also take into account the order's side? For example, should we allow the sell order of A->B and the buy order of B->A from different winners?

Reply:
Side is irrelevant. Solution is a winner if all tokens that appear in its trades weren't used in previous winners.

Reply:
> if all tokens that appear in its trades weren't used in previous winners.

Shouldn't we use `HashSet<TokenAddress>` instead of `HashSet<(TokenAddress, TokenAddress)>` then?

Reply:
switched to flat map, now it's better